byterun/misc.c

```diff
@@ -195,6 +195,50 @@ CAMLexport char * caml_strconcat(int n, ...)
   return res;
 }
 
+/* Integer arithmetic with overflow detection */ 
+
+#if ! (__GNUC__ >= 5 || __has_builtin(__builtin_mul_overflow))
+CAMLexport int caml_mulu_overflow(uintnat a, uintnat b, uintnat * res)
+{
+#define HALF_SIZE (sizeof(uintnat) * 4)
+#define HALF_MASK (((uintnat)1 << HALF_SIZE) - 1)
+#define LOW_HALF(x) ((x) & HALF_MASK)
+#define HIGH_HALF(x) ((x) >> HALF_SIZE)
+  /* Cut in half words */
+  uintnat al = LOW_HALF(a);
+  uintnat ah = HIGH_HALF(a);
+  uintnat bl = LOW_HALF(b);
+  uintnat bh = HIGH_HALF(b);
+  /* Exact product is:
+              al * bl
+           +  ah * bl  << HALF_SIZE
+           +  al * bh  << HALF_SIZE
+           +  ah * bh  << 2*HALF_SIZE
+     Overflow occurs if:
+        ah * bh is not 0, i.e. ah != 0 and bh != 0
+     OR ah * bl has high half != 0
+     OR al * bh has high half != 0
+     OR the sum al * bl + LOW_HALF(ah * bl) << HALF_SIZE
+                        + LOW_HALF(al * bh) << HALF_SIZE overflows.
+     This sum is equal to p = (a * b) modulo word size. */
```

Comment:
The following block can be rewritten so that only a single multiplication is performed in the common case:
``` c
uintnat p = a * b;
*res = p;
if (ah == 0 && bh == 0) return 0;
if (ah != 0 && bh != 0) return 1;
uintnat p1 = al * bh;
uintnat p2 = ah * bl;
if (HIGH_HALF(p1) != 0 || HIGH_HALF(p2) != 0) return 1;
```

This benchmark becomes 10% faster on my PC (modern i5, GCC 5.4, -O2):
```c
#include <stdio.h>
#include <stdint.h>
#include <limits.h>

#define LONG_BIT 64
typedef uint64_t uintnat;

static inline int caml_mulu_overflow(uintnat a, uintnat b, uintnat *res) {
   // ...
}

// Performing a multiplication at all bit sizes,
// assuming that sizes are logarithmically distributed.
int main()
{
  unsigned long i, j, k, mult, total = 0;
  int overflow_count = 0;
  for (k = 0; k < 0x40000000 / LONG_BIT / LONG_BIT; ++k)
    for (i = 0; i != LONG_MAX; i = i*2+1)
      for (j = 0; j != LONG_MAX; j = j*2+1) {
        overflow_count += caml_mulu_overflow(i+k, j+k, &mult);
        total += mult;
      }
  printf("%d overflows (total %lu)\n", overflow_count, total);
}
```